Tensile/SolutionStructs.py

```diff
@@ -3684,6 +3662,163 @@ def assignDerivedParameters(state):
+              optPadMN_Minus = (extraPadDU - optPadMN) % (128//numBytes)
+              optPadMN = min(optPadMN_Plus, optPadMN_Minus)
+            state["LdsPad%s"%tc] = optPadMN
+        ## turn-off padding for directToLds
+        if state["DirectToLds%s"%tc]:
+          state["LdsPad%s"%tc] = 0
+
+      else:
+        # non MI case
+        if state["UnrollMajorLDS%s"%tc]:
+          reject(state, "didn't support UnrollMajorLDS in VALU mode yet")
+        if state["LdsBlockSizePerPad%s"%tc] != 0:
+          reject(state, "didn't support LdsBlockSizePerPad in VALU mode yet")
+
+      assert(state["LdsPad%s"%tc] >= 0)
+
+      # set LdsBlockSizePerPad = 0 if LdsPad is 0
+      if state["LdsPad%s"%tc] == 0:
+        state["LdsBlockSizePerPad%s"%tc] = 0
+
+    # get number of elements in LDS for each tc (A or B)
+    def getLdsNumElements(tc):
+      bpeAB = int(4*state["ProblemType"]["DataType"].numRegisters())
+      if state["LdsBlockSizePerPad%s"%tc]:
+        padInterval = state["LdsBlockSizePerPad%s"%tc] // bpeAB
+      else:
+        if state["UnrollMajorLDS%s"%tc]:
+          padInterval = state["_DepthULds"] 
+        else:
+          padInterval = state["MacroTile%s"%tc]
+      numElements = state["MacroTile%s"%tc] * state["_DepthULds"]
+      ldsNumElements = numElements + (numElements // padInterval) * (state["LdsPad%s"%tc])
+      if state["DirectToVgpr%s"%tc]:
+        # DirectToVgpr does not use LDS. Set to 0.
+        ldsNumElements = 0
+      return ldsNumElements
+
+    # get number of aligned elements in LDS for each tc (A or B)
+    def getLdsNumElementsAligned(tc):
+      ldsAlign = int(64 / state["ProblemType"]["DataType"].numRegisters())
+      ldsNumElements = getLdsNumElements(tc)
+      ldsNumElementsAligned = roundUpToNearestMultiple(ldsNumElements, ldsAlign)
+      return ldsNumElementsAligned
+
+    # LDS size check for auto adjustment
+    # if LDS size is over the limit, change LdsPad to 0
+    # 1. check UnrollMajorLDS=False first
+    for tc in ('A','B'):
+      ldsNumElementsAlignedAB = getLdsNumElementsAligned('A') + getLdsNumElementsAligned('B')
+      if autoAdjusted["LdsPad%s"%tc] and ldsNumElementsAlignedAB > globalParameters["MaxLDS"]:
+        # auto adjusted LdsPad and LDS overflow
+        if (not state["UnrollMajorLDS%s"%tc]):
+          # change LdsPad and LdsBlockSizePerPad to 0
+          state["LdsPad%s"%tc] = 0
+          state["LdsBlockSizePerPad%s"%tc] = 0
+    # 2. check UnrollMajorLDS=True
+    for tc in ('A','B'):
+      ldsNumElementsAlignedAB = getLdsNumElementsAligned('A') + getLdsNumElementsAligned('B')
+      if autoAdjusted["LdsPad%s"%tc] and ldsNumElementsAlignedAB > globalParameters["MaxLDS"]:
+        # auto adjusted LdsPad and LDS overflow
+        if state["UnrollMajorLDS%s"%tc]:
+          # change LdsPad to 0
+          # change LdsPad and LdsBlockSizePerPad to 0
+          state["LdsPad%s"%tc] = 0
+          state["LdsBlockSizePerPad%s"%tc] = 0
```

Comment:
These loops are nearly identical, can they be combined?

Reply:
OK I can merge them.
Like this

# check UnrollMajorLDS=False for A, B first, then true
for umlds in [False, True]:
    for tc in ('A','B'):
        ....
        if state["UnrollMajorLDS%s"%tc] == umlds:
            .....

Reply:
Updated. Merged them.